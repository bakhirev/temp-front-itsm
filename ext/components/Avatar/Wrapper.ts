import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';
import type { ITheme } from '../themes';
import { Subtitle1, Body1Short, Caption1 } from '../Typography';

export type Size = 56 | 48 | 40 | 32 | 24;
export type Kind =
  | 'white'
  | 'light'
  | 'grey'
  | 'dark'
  | 'blue'
  | 'green'
  | 'red'
  | 'orange'
  | 'purple'
  | 'yellow';

const MARGIN = {
  56: '12px',
  48: '10px',
  40: '8px',
  32: '6px',
  24: '4px',
};

const getBackgroundColor = (kind: Kind, theme: ITheme) => {
  switch (kind) {
    case 'white':
      return theme.color.neutral.white;
    case 'light':
      return theme.color.neutral[10];
    case 'grey':
      return theme.color.neutral[50];
    case 'dark':
      return theme.color.neutral[80];
    case 'blue':
      return theme.color.primary[20];
    case 'green':
      return theme.color.success[20];
    case 'red':
      return theme.color.error[20];
    case 'orange':
      return theme.color.warning[20];
    case 'purple':
      return theme.color.extraOne[20];
    case 'yellow':
      return theme.color.attention[20];
  }
};

const getContentColor = (kind: Kind, theme: ITheme) => {
  if (kind === 'grey' || kind === 'dark') return theme.color.neutral.white;
  return theme.color.neutral[90];
};

const commonWrapperStyles = (size: Size, kind: Kind, theme: ITheme) => css`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  width: ${size}px;
  height: ${size}px;
  border-radius: 50%;
  background-color: ${getBackgroundColor(kind, theme)};
  color: ${getContentColor(kind, theme)};
  text-transform: uppercase;
  user-select: none;
  overflow: hidden;

  img,
  svg {
    width: 100%;
    height: 100%;
  }

  svg {
    fill: ${getContentColor(kind, theme)};
    margin: ${MARGIN[size]};
  }
`;

interface IWrapperProps {
  size: Size;
  kind: Kind;
}

const BigWrapper = styled(Subtitle1)<IWrapperProps>`
  ${({ size, kind, theme }) => commonWrapperStyles(size, kind, theme)};
`;

BigWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};

const MediumWrapper = styled(Body1Short)<IWrapperProps>`
  ${({ size, kind, theme }) => commonWrapperStyles(size, kind, theme)};
`;

MediumWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};

const SmallWrapper = styled(Caption1)<IWrapperProps>`
  ${({ size, kind, theme }) => commonWrapperStyles(size, kind, theme)};
`;

SmallWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};

export const Wrapper = (size: Size) => {
  switch (size) {
    case 56:
      return BigWrapper;
    case 48:
    case 40:
      return MediumWrapper;
    case 32:
    case 24:
      return SmallWrapper;
  }
};
