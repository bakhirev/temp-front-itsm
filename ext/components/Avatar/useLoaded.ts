import { useState, useEffect } from 'react';

export const useLoaded = ({ src, srcSet }) => {
  const [loaded, setLoaded] = useState<'loaded' | 'error' | boolean>(false);

  useEffect(() => {
    if (!src && !srcSet) return;
    setLoaded(false);

    let effectActive = true;
    const image = new Image();
    image.src = src;
    if (srcSet) {
      image.srcset = srcSet;
    }

    image.onload = () => {
      if (!effectActive) {
        return;
      }
      setLoaded('loaded');
    };
    image.onerror = () => {
      if (!effectActive) {
        return;
      }
      setLoaded('error');
    };
    return () => {
      effectActive = false;
    };
  }, [src, srcSet]);

  return loaded;
};
