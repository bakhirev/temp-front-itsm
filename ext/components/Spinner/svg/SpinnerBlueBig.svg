<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9922C48.0193 25.3483 46.6569 24 45 24C43.3432 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5067 32.9234 42 24 42C14.0589 42 6.00001 33.9411 6.00001 24C6.00001 15.0766 12.4933 7.6698 21.0126 6.24674C22.6468 5.97377 24 4.65683 24 2.99997C24 1.34312 22.6517 -0.0193537 21.0077 0.187219C17.3213 0.650452 13.7783 1.96536 10.6663 4.0447C6.71955 6.68186 3.64341 10.4301 1.8269 14.8156C0.0103989 19.201 -0.464881 24.0266 0.461164 28.6821C1.38721 33.3377 3.67299 37.6141 7.02945 40.9705C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2804 43.9553 37.3337Z" fill="#0062FF"/>
</svg>
