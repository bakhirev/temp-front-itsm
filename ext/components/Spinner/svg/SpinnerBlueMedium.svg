<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M21.9776 18.6669C23.0173 17.1109 23.6748 15.3394 23.9064 13.4962C24.0097 12.6742 23.3284 12 22.5 12C21.6716 12 21.0131 12.6766 20.8766 13.4937C20.1651 17.7534 16.4617 21 12 21C7.02944 21 3.00001 16.9706 3.00001 12C3.00001 7.53833 6.24663 3.83493 10.5063 3.1234C11.3234 2.98691 12 2.32844 12 1.50002C12 0.671589 11.3258 -0.00964632 10.5039 0.0936401C8.66065 0.325257 6.88914 0.982712 5.33316 2.02238C3.35977 3.34096 1.8217 5.2151 0.913451 7.40782C0.00519943 9.60053 -0.232441 12.0133 0.230582 14.3411C0.693605 16.6689 1.83649 18.8071 3.51472 20.4853C5.19295 22.1635 7.33115 23.3064 9.65892 23.7694C11.9867 24.2325 14.3995 23.9948 16.5922 23.0866C18.7849 22.1783 20.6591 20.6402 21.9776 18.6669Z" fill="#0062FF"/>
</svg>
