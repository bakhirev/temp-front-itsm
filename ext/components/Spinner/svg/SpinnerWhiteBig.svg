<svg width="48" height="48" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M43.9553 37.3337C46.0346 34.2217 47.3495 30.6787 47.8128 26.9923C48.0193 25.3483 46.6569 24 45 24C43.3431 24 42.0262 25.3532 41.7532 26.9874C40.3302 35.5068 32.9234 42 24 42C14.0589 42 6 33.9411 6 24C6 15.0766 12.4932 7.66983 21.0126 6.24677C22.6468 5.9738 24 4.65685 24 3C24 1.34315 22.6517 -0.0193258 21.0077 0.187247C17.3213 0.65048 13.7783 1.96539 10.6663 4.04473C6.71953 6.68188 3.64339 10.4302 1.82689 14.8156C0.0103881 19.201 -0.464892 24.0266 0.461153 28.6822C1.3872 33.3377 3.67298 37.6141 7.02944 40.9706C10.3859 44.327 14.6623 46.6128 19.3178 47.5388C23.9734 48.4649 28.799 47.9896 33.1844 46.1731C37.5698 44.3566 41.3181 41.2805 43.9553 37.3337Z" fill="white"/>
</svg>
