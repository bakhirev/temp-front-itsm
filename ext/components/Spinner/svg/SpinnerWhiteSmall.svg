<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M14.6518 12.4446C15.3449 11.4072 15.7832 10.2262 15.9376 8.99742C16.0064 8.44945 15.5523 8 15 8C14.4477 8 14.0087 8.45108 13.9177 8.99581C13.4434 11.8356 10.9745 14 8 14C4.68629 14 2 11.3137 2 8C2 5.02554 4.16441 2.55661 7.00419 2.08226C7.54893 1.99127 8 1.55228 8 1C8 0.447715 7.55055 -0.00644196 7.00258 0.0624156C5.77376 0.216827 4.59275 0.65513 3.55544 1.34824C2.23985 2.22729 1.21446 3.47672 0.608964 4.93853C0.00346269 6.40034 -0.154964 8.00887 0.153718 9.56072C0.4624 11.1126 1.22433 12.538 2.34315 13.6569C3.46197 14.7757 4.88743 15.5376 6.43928 15.8463C7.99113 16.155 9.59966 15.9965 11.0615 15.391C12.5233 14.7855 13.7727 13.7602 14.6518 12.4446Z" fill="white"/>
</svg>
