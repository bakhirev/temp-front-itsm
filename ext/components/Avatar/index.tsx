import React, { FC, ReactNode } from 'react';

import { ReactComponent as PersonOutline } from '../Icons/system/PersonOutline.svg';

import type { Size, Kind } from './Wrapper';
import { Wrapper } from './Wrapper';
import { useLoaded } from './useLoaded';

export type IAvatarProps = {
  /** src атрибут для img элемента. Также в компонент можно передать icon или text элементы в качестве children,
   * children будут отрисованы, если src не установлен или загрузка img завершилась ошибкой */
  src?: string;
  /** srcSet атрибут для img элемента */
  srcSet?: string;
  /** sizes атрибут для img элемента */
  sizes?: string;
  /** alt атрибут для img элемента */
  alt?: string;
  /** Тип компонента */
  kind?: Kind;
  /** Размер компонента */
  size?: Size;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
};

export const Avatar: FC<IAvatarProps> = ({
  className,
  dataTestId,
  kind = 'white',
  size = 56,
  src,
  srcSet,
  sizes,
  alt,
  children,
}) => {
  const hasImg = src || srcSet;
  const loaded = useLoaded({ src, srcSet });
  const hasImgLoaded = hasImg && loaded === 'loaded';

  const getInitials = (alt: string) => {
    const words = alt.split(' ');
    return words.length > 1 ? words[0][0] + words[1][0] : words[0][0];
  };
  const controlStringLength = (str: string) => {
    return str.length > 2 ? str.substr(0, 2) : str;
  };

  const getContent = () => {
    let content: ReactNode = null;
    if (hasImgLoaded) {
      content = <img alt={alt} src={src} srcSet={srcSet} sizes={sizes} />;
    } else if (children) {
      content = typeof children === 'string' ? controlStringLength(children) : children;
    } else if (hasImg && alt) {
      content = getInitials(alt);
    } else {
      content = <PersonOutline />;
    }
    return content;
  };

  const AvatarWrapper = Wrapper(size);

  return (
    <>
      <AvatarWrapper className={className} data-test-id={dataTestId} size={size} kind={kind}>
        {getContent()}
      </AvatarWrapper>
    </>
  );
};
