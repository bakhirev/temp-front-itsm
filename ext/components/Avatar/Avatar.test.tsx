import React from 'react';
import { shallow, mount } from 'enzyme';

import { Avatar } from '../Avatar';
import { ReactComponent as PersonOutlineIcon } from '../Icons/system/PersonOutlineIcon.svg';
import { ReactComponent as AddpersonOutlineIcon } from '../Icons/system/AddpersonOutlineIcon.svg';

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    color: {
      neutral: {
        white: '',
      },
    },
  },
}));

describe('Avatar', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const validSrc = 'https://duckduckgo.com/assets/common/dax-logo.svg';
  const invalidSrc = 'https://duckdusets/common/dax-logo.svg';
  const alt = 'Big Duck';
  const text = 'ABCD';

  it('should render component with default kind, size and icon', () => {
    const wrapper = shallow(<Avatar />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.childAt(0).is(PersonOutlineIcon));
  });

  it('should render component with text no more than two characters long', () => {
    const wrapper = shallow(<Avatar>{text}</Avatar>);
    expect(wrapper.childAt(0).render().text()).toBe('AB');
  });

  it('should render component with kind Blue and size 32', () => {
    const wrapper = shallow(<Avatar kind="blue" size={32} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with IMAGE', () => {
    const wrapper = mount(<Avatar src={validSrc} alt={alt} />);
    setTimeout(() => {
      expect(wrapper.find('img').prop('src')).toEqual(validSrc);
    }, 10000);
  });

  it('should render component with initials taken from alt attribute if image could not be loaded ', () => {
    const wrapper = shallow(<Avatar src={invalidSrc} alt={alt} />);
    expect(wrapper.childAt(0).render().text()).toBe('BD');
  });

  it('should render component with icon', () => {
    const wrapper = shallow(
      <Avatar>
        <AddpersonOutlineIcon />
      </Avatar>
    );
    expect(wrapper.childAt(0).is(AddpersonOutlineIcon));
  });
});
