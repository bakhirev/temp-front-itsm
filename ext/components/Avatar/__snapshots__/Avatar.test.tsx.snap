// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render component with default kind, size and icon 1`] = `
<Fragment>
  <Styled(styled.div)
    kind="white"
    size={56}
    theme={
      Object {
        "color": Object {
          "neutral": Object {
            "white": "",
          },
        },
      }
    }
  >
    <div />
  </Styled(styled.div)>
</Fragment>
`;

exports[`Avatar should render component with kind Blue and size 32 1`] = `
<Fragment>
  <Styled(styled.div)
    kind="blue"
    size={32}
    theme={
      Object {
        "color": Object {
          "neutral": Object {
            "white": "",
          },
        },
      }
    }
  >
    <div />
  </Styled(styled.div)>
</Fragment>
`;
